ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- TODO: this module should be exposed as 'Ouorboros.Network'
+module Ouroboros.Network.Mux.Interface
+  ( NetworkInterface (..)
+  , MuxPeer (..)
+  , NetworkNode (..)
+
+  -- * Auxiliary functions
+  , miniProtocolDescription
+  ) where
+
+import qualified Codec.CBOR.Read     as CBOR
+import           Data.ByteString.Lazy (ByteString)
+import           Data.Functor (void)
+import           Numeric.Natural (Natural)
+
+import           Control.Monad.Class.MonadAsync ( MonadAsync )
+import           Control.Monad.Class.MonadThrow ( MonadCatch
+                                                , MonadThrow
+                                                )
+
+import           Network.TypedProtocol.Core
+import           Network.TypedProtocol.Codec
+import           Network.TypedProtocol.Driver
+import           Network.TypedProtocol.Pipelined
+
+import           Ouroboros.Network.Mux.Types
+
+-- |
+-- Specification for peers of a protocol.  This type instructs the multiplexing
+-- layer to run a given client \/ server peers.
+--
+data MuxPeer m where
+  -- |
+  -- A non pipeliend peer together with a codec.
+  --
+  OnlyClient
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsClient st m a
+    -> MuxPeer m
+
+  -- |
+  -- A pipelined peer together with a codec.
+  --
+  OnlyPipelinedClient
+    :: forall ps (st :: ps) m a.
+       Natural
+    -> Codec ps CBOR.DeserialiseFailure m ByteString
+    -> PeerPipelined ps AsClient st m a
+    -> MuxPeer m
+
+  -- |
+  -- Server peer with a codec
+  OnlyServer
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+  -- |
+  -- Client and server peers with the corresponding codec.
+  --
+  ClientAndServer
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsClient st m a
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+  -- |
+  -- Pipelined client and a server with the correspnding codec.
+  --
+  PipelinedClientAndServer
+    :: forall ps (st :: ps) m a.
+       Natural
+    -> Codec ps CBOR.DeserialiseFailure m ByteString
+    -> PeerPipelined ps AsClient st m a
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+
+-- |
+-- Public network interface for 'ouroboros-network'.
+--
+data NetworkInterface ptcl addr m = NetworkInterface {
+      -- |
+      -- Address of the node to run.  The node will bind to this address, and
+      -- listen for incoming connections.  Some bearers do not have a notion of
+      -- address.
+      --
+      nodeAddress :: addr, 
+
+      -- |
+      -- Versioned set of peers to run; it is the responsibilty of the caller
+      -- to provide the correct associtation between `ptcl` and a protocol
+      -- (which is existentially quantified using @'PeerSpecification'@).
+      --
+      protocols   :: [(Version, ptcl -> MuxPeer m)]
```

Comment:
We may later need to locally existentially quantify over the `ptcl`, if we end up with different sets for different versions.

Reply:
Or to make `Version` depend on `ptcl` and have special tuple to combine the right version with the right `ptcl`:
```
data Version ptcl where
  Version0 :: Version Protocols0
  Version1 :: Version Protocols0
  Version2 :: Version Protocols1

data VersionAndPTCL where
  VerssionAndPTCL :: Version ptcl -> ptcl -> VersionAndPTCL
```